### PR TITLE
bpo-44525: Specialize simple Python calls.

### PR DESCRIPTION
**Pull request by markshannon**
Complement to https://github.com/python/cpython/pull/26934
[About 1% speedup](https://gist.github.com/markshannon/bf0d916db3814bb0545ff94bdc19e20f)

https://bugs.python.org/issue44525

### CODE REVIEW

File: Python/specialize.c [hunk 1]
```diff
@@ -1352,9 +1413,15 @@ _Py_Specialize_CallFunction(
...
+    else if (PyFunction_Check(callable)) {
+        fail = specialize_py_call((PyFunctionObject *)callable, instr, nargs, cache);
+    }
+    else if (PyType_Check(callable)) {
+        fail = specialize_class_call(callable, instr, nargs, cache);
+    }
     else {
         SPECIALIZATION_FAIL(CALL_FUNCTION, call_fail_kind(callable));
-        fail = 1;
+        fail = -1;
```
Comment (Fidget-Spinner) [inline Python/specialize.c:1424]: I set `specialize_c_call` to return 1 to indicate failure. You may have to change the return codes in  `specialize_c_call` for consistency. Sorry!

Comment (markshannon) [reply Python/specialize.c:1424]: We should probably add an enum `{SUCCESS, FAILURE}` for clarity.

I wouldn't worry too much about it for now, we do much worse elsewhere.
`compile.c` uses zero for success and non-zero for failure in some places, and the other way around in other places :disappointed:

File: Python/ceval.c [hunk 2]
```diff
@@ -4732,9 +4732,48 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            if (new_frame == NULL) {
+                goto error;
+            }
+            STACK_SHRINK(argcount);
+            for (int i = 0; i < argcount; i++) {
+                new_frame->localsplus[i] = stack_pointer[i];
+            }
+            int deflen = cache1->defaults_len;
+            for (int i = 0; i < deflen; i++) {
+                PyObject *def = PyTuple_GET_ITEM(func->func_defaults, cache1->defaults_start+i);
```
Comment (Fidget-Spinner) [inline Python/ceval.c:4766]: I wonder if the compiler is smart enough to realize that `func->func_defaults` and `cache1->defaults_start` doesn't change between iterations? I'd say probably not. Perhaps we should micro-optimize and move this out of the loop?

Comment (markshannon) [reply Python/ceval.c:4766]: The common case is that `deflen == 0`. In which case moving it out of loop could make it worse.

**Comment by markshannon** [comment]
The usual buildbot failures.

**Pull request merged**